### PR TITLE
node-bootstrapper

### PR DESCRIPTION
**Pull request by r2k1**
The current implementation is a simple wrapper around CSE scripts and a custom data generator for Linux (previously known as "self-contained" efforts).

This approach moves the script generation logic from the API layer into the VHD itself, making virtual machines more consistent and allowing for significant simplification.

By having a Go application as the main entry point to provision the VHD, we unlock a lot of new possibilities.

Switching to the new generation should be as simple as replacing `GetNodeBootstrapping` with `GetNodeBootstrappingV2`.

No changes are introduced to the current VHD provisioning.

### CODE REVIEW

File: installer/build.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+GOOS=linux GOARCH=amd64 go build -o ./dist/installer-linux-amd64
```
Comment (smith1511) [inline installer/build.sh:1]: Maybe prefix with "node-..." to make the binaries more self-descriptive?

File: pkg/agent/bakerapi.go [hunk 2]
```diff
@@ -85,6 +87,20 @@ func (agentBaker *agentBakerImpl) GetNodeBootstrapping(ctx context.Context, conf
 	return nodeBootstrapping, nil
 }
 
+func (agentBaker *agentBakerImpl) GetNodeBootstrappingV2(ctx context.Context, config *datamodel.NodeBootstrappingConfiguration) (*datamodel.NodeBootstrapping, error) {
```
Comment (smith1511) [inline pkg/agent/bakerapi.go:90]: GetNodeBootstrappingForSelfContained()?

**Comment by Devinwong** [comment]
Who will trigger the main.go and providing the nodebootstrapping config? Assuming it's baked in the VHD, right?

**Comment by r2k1** [comment]
The binary is compiled and added to every new linux VHD.

AgentBaker Server is preparing and running this CSE command:
```
fmt.Sprintf(
		`bash -c "(echo '%s' | base64 -d > config.json && mkdir -p /var/log/azure && /opt/azure/node-bootstrapper provision --provision-config=config.json) > /var/log/azure/node-bootstrapper.log 2>&1"`,
		base64.StdEncoding.EncodeToString(configJSON),
	)
```

We may change it the future. My goal was to exactly replicate current functionality with minimum changes.

In the future we may change this behaviour to a daemon watching for a config file.

File: e2e/validators.go [hunk 3]
```diff
@@ -269,6 +269,21 @@ func containerdVersionValidator(version string) *LiveVMValidator {
 	}
 }
 
+func textValidator(path string, expected string) *LiveVMValidator {
```
Comment (cameronmeissner) [inline e2e/validators.go:272]: doesn't `FileHasContentsValidator` accomplish the same thing?

File: e2e/nodebootstrapping.go [hunk 4]
```diff
@@ -52,11 +52,11 @@ func getNodeBootstrappingForValidation(ctx context.Context, nbc *datamodel.NodeB
 	if err != nil {
 		return nil, err
 	}
-	nodeBootstrapping, err := ab.GetNodeBootstrapping(ctx, nbc)
-	if err != nil {
-		return nil, err
+	if config.Config.EnableNodeBootstrapperTest {
+		return ab.GetNodeBootstrappingV2(ctx, nbc)
+	} else {
```
Comment (cameronmeissner) [inline e2e/nodebootstrapping.go:57]: nit: redundant else

File: e2e/scenario_test.go [hunk 5]
```diff
@@ -892,6 +892,7 @@ func Test_ubuntu2204gpuNoDriver(t *testing.T) {
 }
 
 func Test_ubuntu2204privatekubepkg(t *testing.T) {
+	t.Skip("this VHD can't be updated and self-contained logic doesn't work for it")
```
Comment (cameronmeissner) [inline e2e/scenario_test.go:895]: we probably don't want to blindly disable this, would there be a way we could keep it around for non self-contained?

File: pkg/agent/baker.go [hunk 6]
```diff
@@ -1001,6 +979,44 @@ func getContainerServiceFuncMap(config *datamodel.NodeBootstrappingConfiguration
...
+		profile.KubernetesConfig.ContainerRuntimeConfig[datamodel.ContainerDataDirKey] != "" {
+		return profile.KubernetesConfig.ContainerRuntimeConfig[datamodel.ContainerDataDirKey]
+	}
+	if profile.KubeletDiskType == datamodel.TempDisk {
+		return datamodel.TempDiskContainerDataDir
+	}
+	return cs.Properties.OrchestratorProfile.KubernetesConfig.ContainerRuntimeConfig[datamodel.ContainerDataDirKey]
+}
+
+func HasDataDir(config *datamodel.NodeBootstrappingConfiguration) bool {
```
Comment (cameronmeissner) [inline pkg/agent/baker.go:996]: could we add some UTs for these while we're at it?

File: pkg/agent/datamodel/types.go [hunk 7]
```diff
@@ -1742,6 +1742,9 @@ type NodeBootstrappingConfiguration struct {
 	// CNI, which will overwrite the `filter` table so that we can only insert to `mangle` table to avoid
 	// our added rule is overwritten by Cilium.
 	InsertIMDSRestrictionRuleToMangleTable bool
+
+	// Version is require for node-bootstrapper application to determine the version of the config file.
```
Comment (cameronmeissner) [inline pkg/agent/datamodel/types.go:1746]:
```suggestion
	// Version is required for node-bootstrapper application to determine the version of the config file.
```

File: node-bootstrapper/build.sh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
```
Comment (cameronmeissner) [inline node-bootstrapper/build.sh:1]: fine for now, though might be worth considering a Makefile

File: node-bootstrapper/main.go [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+		os.Exit(1)
+	}
+	slog.Info("node-bootstrapper finished")
+}
+
+func Run(ctx context.Context) error {
+	if len(os.Args) < 2 {
+		return errors.New("missing command argument")
+	}
+	switch os.Args[1] {
```
Comment (cameronmeissner) [inline node-bootstrapper/main.go:44]: we could probably use something like `cobra` to accomplish this

**Comment by cameronmeissner** [comment]
will take another closer look tomorrow

File: e2e/validators.go (hunk 3 above)
Comment (r2k1) [reply e2e/validators.go:272]: I don't think I indented to touch this part. Probably needs to be rebased with master.

File: e2e/scenario_test.go (hunk 5 above)
Comment (r2k1) [reply e2e/scenario_test.go:895]: Can be done with adding `if config.EnableNodeBootstrapperTest`

File: pkg/agent/baker.go (hunk 6 above)
Comment (r2k1) [reply pkg/agent/baker.go:996]: I didn't really touch anything there. 

I exported a few private methods. And did a few "right-click => extract method". So, I can share exactly the same logic of file templating with node-bootstrapper.

File: installer/build.sh (hunk 1 above)
Comment (timmy-wright) [reply installer/build.sh]: Changed.

File: pkg/agent/bakerapi.go (hunk 2 above)
Comment (timmy-wright) [reply pkg/agent/bakerapi.go]: Changed.

File: e2e/nodebootstrapping.go (hunk 4 above)
Comment (timmy-wright) [reply e2e/nodebootstrapping.go]: Removed it.

File: pkg/agent/datamodel/types.go (hunk 7 above)
Comment (timmy-wright) [reply pkg/agent/datamodel/types.go]: Fixed.

File: e2e/validators.go (hunk 3 above)
Comment (timmy-wright) [reply e2e/validators.go:268]: Have removed this and replaced with FileHasContentsValidator

File: pkg/agent/baker.go (hunk 6 above)
Comment (timmy-wright) [reply pkg/agent/baker.go:996]: Have added tests for extracted methods.

File: node-bootstrapper/build.sh (hunk 8 above)
Comment (timmy-wright) [reply node-bootstrapper/build.sh:1]: I agree - using the same build system as the rest of AgentBaker is important. Will change this in a later PR. I have got the testing and linting working with the makefile tho.

**Comment by Bryce-Soghigian** [comment]
👀

File: e2e/scenario_test.go (hunk 5 above)
Comment (timmy-wright) [reply e2e/scenario_test.go]: Have removed this skip and added a way for tests to specify which bootstrapping method they want.

File: self-contained/bootstrap_install.sh [hunk 10]
```diff
@@ -60,6 +60,9 @@ installContainerRuntime() {
 }
 
 installNetworkPlugin() {
+    if [[ "${NETWORK_PLUGIN}" = "none" ]]; then
+        return
+    fi
```
Comment (r2k1) [inline self-contained/bootstrap_install.sh:65]: @timmy-wright did you change it intentionally?

File: pkg/agent/bakerapi.go [hunk 11]
```diff
@@ -85,6 +87,32 @@ func (agentBaker *agentBakerImpl) GetNodeBootstrapping(ctx context.Context, conf
...
+	config.Version = "v0"
+	nodeBootstrapping, err := agentBaker.GetNodeBootstrapping(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+	configJSON, err := json.Marshal(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal nbc, error: %w", err)
+	}
+	//nolint: lll
```
Comment (djsly) [inline pkg/agent/bakerapi.go:107]: what is `lll`

File: node-bootstrapper/custom_data_test.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+		actual := getFile(t, nbc, "/etc/kubernetes/certs/ca.crt", 0600)
+		expected := "test-ca-cert"
+		assert.Equal(t, expected, actual)
+	})
+
+	t.Run("exec_start.conf", func(t *testing.T) {
+		nbc := validNBC()
+		actual := getFile(t, nbc, "/etc/systemd/system/docker.service.d/exec_start.conf", 0644)
+		nbc.ContainerService.Properties.OrchestratorProfile.KubernetesConfig.DockerBridgeSubnet = "1.1.1.1"
+		expected := `[Service]
```
Comment (djsly) [inline node-bootstrapper/custom_data_test.go:22]: where is the expected custom data coming from ?  

do we need to update this testdata when we change the content of the files ?

File: node-bootstrapper/main.go [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+		return fmt.Errorf("content bootstrap kubeconfig: %w", err)
+	}
+	files["/var/lib/kubelet/bootstrap-kubeconfig"] = File{
+		Content: bootstrapKubeconfig,
+		Mode:    ReadOnlyWorld,
+	}
+	return nil
+}
+
+func contentKubeconfig(config *datamodel.NodeBootstrappingConfiguration) string {
```
Comment (djsly) [inline node-bootstrapper/main.go:236]: could we renamed all function starting with `content<artifact>` with maybe something like `generate<artifact>Content` 

When reading the function name I wasn't sure what was happening.

File: node-bootstrapper/main.go [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+	users := `- name: client
+  user:
+    client-certificate: /etc/kubernetes/certs/client.crt
+    client-key: /etc/kubernetes/certs/client.key`
+
+	return fmt.Sprintf(`
```
Comment (djsly) [inline node-bootstrapper/main.go:242]: I feel generating inline string with explicit format is asking for trouble. could we look instead at migrating those template string outside of go and load template files instead ?

File: node-bootstrapper/main.go [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+- context:
+    cluster: localcluster
+    user: client
+  name: localclustercontext
+current-context: localclustercontext
+`, agent.GetKubernetesEndpoint(config.ContainerService), users)
+}
+
+func contentBootstrapKubeconfig(config *datamodel.NodeBootstrappingConfiguration) (string, error) {
+	data := map[string]any{
```
Comment (djsly) [inline node-bootstrapper/main.go:262]: for example, the generated string is really hard to grasp from the code. Having an external template file would help quickly identify what the final file content looks like (formatting too) and which parameters are coming from outside.

File: node-bootstrapper/build.sh (hunk 8 above)
Comment (djsly) [reply node-bootstrapper/build.sh:1]: agree as well.  let align on build process

File: e2e/validation.go [hunk 16]
```diff
@@ -110,7 +110,7 @@ func commonLiveVMValidators(opts *scenarioRunOpts) []*LiveVMValidator {
 				"cluster-provision-cse-output.log",
 				"cloud-init-files.paved",
 				"vhd-install.complete",
-				"cloud-config.txt",
+				//"cloud-config.txt", // file with UserData
```
Comment (djsly) [inline e2e/validation.go:113]: left over TODO ? no sure why this needs to be removed

File: e2e/scenario_test.go [hunk 17]
```diff
@@ -559,6 +559,20 @@ func Test_ubuntu1804ChronyRestarts(t *testing.T) {
 	})
 }
 
+func Test_ubuntu2204SelfContainedInstaller(t *testing.T) {
+	RunScenario(t, &Scenario{
+		Description: "tests that a new ubuntu 2204 node using self contained installer can be properly bootstrapped",
+		Config: Config{
+			NodeBootstrappingType: SelfContained,
```
Comment (djsly) [inline e2e/scenario_test.go:566]: do we want to use that type `selfcontained` ?  I know we moved to calling this `scriptless` movgin forward. 

What are other expected types ?

File: e2e/nodebootstrapping.go [hunk 18]
```diff
@@ -47,16 +47,20 @@ func getNodeBootstrappingForCoverage(nbc *datamodel.NodeBootstrappingConfigurati
...
+func getNodeBootstrappingForValidation(ctx context.Context, nbc *datamodel.NodeBootstrappingConfiguration, bootstrappingType NodeBootstrappingType) (*datamodel.NodeBootstrapping, error) {
 	ab, err := agent.NewAgentBaker()
 	if err != nil {
 		return nil, err
 	}
-	nodeBootstrapping, err := ab.GetNodeBootstrapping(ctx, nbc)
-	if err != nil {
-		return nil, err
+	switch {
+	case bootstrappingType == SelfContained:
```
Comment (djsly) [inline e2e/nodebootstrapping.go:56]: should we renamed this to scriptless ?

File: e2e/nodebootstrapping.go [hunk 19]
```diff
@@ -47,16 +47,20 @@ func getNodeBootstrappingForCoverage(nbc *datamodel.NodeBootstrappingConfigurati
...
+		return ab.GetNodeBootstrappingForSelfContained(ctx, nbc)
+	case bootstrappingType == CustomScripts:
```
Comment (djsly) [inline e2e/nodebootstrapping.go:58]: what is customScript ?

File: e2e/installer_test.go [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+		Description: "Tests that a node using the Ubuntu 2204 VHD can be properly bootstrapped",
+		Config: Config{
+			//NodeBootstrappingType: SelfContained,
+			Cluster: ClusterKubenet,
+			VHD:     config.VHDUbuntu2204Gen2Containerd,
+			LiveVMValidators: []*LiveVMValidator{
+				containerdVersionValidator("1.7.20-1"),
+				runcVersionValidator("1.1.12-1"),
+			},
+			CSEOverride:       CSENodeBootstrapper(context.TODO(), t, cluster),
```
Comment (djsly) [inline e2e/installer_test.go:39]: what is context.TODO() ?

File: e2e/installer_test.go [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+	configJSON, err := json.Marshal(cluster.NodeBootstrappingConfiguration)
+	require.NoError(t, err)
+
+	binary := compileNodeBootstrapper(t)
+	url, err := config.Azure.UploadAndGetLink(ctx, "node-bootstrapper-"+hashFile(t, binary.Name()), binary)
+	require.NoError(t, err)
+	return fmt.Sprintf(`bash -c "(echo '%s' | base64 -d > config.json && curl -L -o ./node-bootstrapper '%s' && chmod +x ./node-bootstrapper && mkdir -p /var/log/azure && ./node-bootstrapper provision --provision-config=config.json) > /var/log/azure/node-bootstrapper.log 2>&1"`, base64.StdEncoding.EncodeToString(configJSON), url)
+}
+
+func compileNodeBootstrapper(t *testing.T) *os.File {
```
Comment (djsly) [inline e2e/installer_test.go:55]: are we recompiling the binary with each test ? should look at making this more efficient

File: e2e/config/config.go [hunk 22]
```diff
@@ -32,6 +32,9 @@ type Configuration struct {
 	IgnoreScenariosWithMissingVHD bool          `env:"IGNORE_SCENARIOS_WITH_MISSING_VHD"`
 	SkipTestsWithSKUCapacityIssue bool          `env:"SKIP_TESTS_WITH_SKU_CAPACITY_ISSUE"`
 	KeepVMSS                      bool          `env:"KEEP_VMSS"`
+	BlobStorageAccount            string        `env:"BLOB_STORAGE_ACCOUNT" envDefault:"https://abe2e.blob.core.windows.net"`
```
Comment (djsly) [inline e2e/config/config.go:35]: is this a new blob storage ? how was this created ? do we have IaC for it? if this is purely to host the binary for the test, I'm wondering if we could avoid that, knowing that the more azure resources we have to more SFI work we have to do.

File: e2e/config/azure.go [hunk 23]
```diff
@@ -126,37 +134,74 @@ func NewAzureClient(subscription string) (*AzureClient, error) {
 
 	cloud.VMSS, err = armcompute.NewVirtualMachineScaleSetsClient(subscription, credential, opts)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create vmss client: %w", err)
```
Comment (djsly) [inline e2e/config/azure.go:129]: nit: fail/failed are keywords that are normally used when grepping for what went wrong. I"m not sure how practical it is to remove them here, I would personally keep them

File: .golangci.yaml [hunk 24]
```diff
@@ -148,7 +148,7 @@ linters-settings:
       shadow:
         # Whether to be strict about shadowing; can be noisy.
         # Default: false
-        strict: true
+        strict: false
```
Comment (djsly) [inline .golangci.yaml:151]: why was this changed ?

File: pkg/agent/bakerapi.go (hunk 11 above)
Comment (r2k1) [reply pkg/agent/bakerapi.go:107]: The line with bash script is too long, but I don't want to split it across multiple lines.

File: pkg/agent/baker.go [hunk 25]
```diff
@@ -1001,6 +979,44 @@ func getContainerServiceFuncMap(config *datamodel.NodeBootstrappingConfiguration
...
 
+func GetDataDir(config *datamodel.NodeBootstrappingConfiguration) string {
+	cs := config.ContainerService
+	profile := config.AgentPoolProfile
+	if profile != nil && profile.KubernetesConfig != nil &&
+		profile.KubernetesConfig.ContainerRuntimeConfig != nil &&
+		profile.KubernetesConfig.ContainerRuntimeConfig[datamodel.ContainerDataDirKey] != "" {
+		return profile.KubernetesConfig.ContainerRuntimeConfig[datamodel.ContainerDataDirKey]
+	}
+	if profile.KubeletDiskType == datamodel.TempDisk {
```
[the remark this answers is not in the document]
Comment (r2k1) [reply pkg/agent/baker.go:990]: I didn't change the code, just extracted it to an exported function I can use from node-bootstrapper.

It will be a panic, but I wanted to minimize any changes outside of node-boostrapper.

File: e2e/installer_test.go (hunk 21 above)
Comment (r2k1) [reply e2e/installer_test.go:55]: Good point, I'll take a look.

Initially, there's been a single test. Definitely should re-use the binary.

File: e2e/config/config.go (hunk 22 above)
Comment (r2k1) [reply e2e/config/config.go:35]: It's a hack to speed e2e feedback loop.

Without it we need to rebuild the VHD before running any tests. It adds about an hour to the loop.
This code put binary on a node and run a similar CSE command to what we're going to use.

I had difficulty finding a way to upload a binary to a VM. It's not available from a public network. So I couldn't SSH into it. And it's not joined k8s cluster yet when we need to execute it.

I think I created it manually (it's been a while ago).

I don't really like and happy to change if you have any other ideas.

File: e2e/validation.go (hunk 16 above)
Comment (r2k1) [reply e2e/validation.go:113]: This file doesn't exist for scriptless approach. Because we don't supply UserData at all. (It mostly contains bash script).

File: node-bootstrapper/main.go (hunk 15 above)
Comment (r2k1) [reply node-bootstrapper/main.go:262]: I hate YAML-generation in templates. Especially when IF conditions are involved.
It gives me HELM vibes with hacks around proper indentation. Tabs vs spaces etc.

In this function I intentionally avoided using templates due to its complexity.

File: node-bootstrapper/custom_data_test.go (hunk 12 above)
Comment (r2k1) [reply node-bootstrapper/custom_data_test.go:22]: There is no "source" for the data. Mainly a reflection of what's currently happening.

Yes, it needs to be updated when files are changed.

File: node-bootstrapper/main.go (hunk 14 above)
Comment (r2k1) [reply node-bootstrapper/main.go:242]: I want to delay the decision on how to handle it properly.

Currently there only a couple of small files and I think it's easier to do this way without an overhead of templating.

But I agree, it's probably not sustainable if we're going to move all files we generate. I'm just not sure yet how to approach it.

File: .golangci.yaml (hunk 24 above)
Comment (r2k1) [reply .golangci.yaml:151]: The linter is being really noisy for a typical go error handling.
I think shadowing errors is fine, we shouldn't come up with err2 and err3 variables.

File: e2e/installer_test.go (hunk 20 above)
Comment (r2k1) [reply e2e/installer_test.go]: Similar to `context.Background`, a context that never gets cancelled.
I replaced to a one with an expiration time.

File: e2e/config/azure.go (hunk 23 above)
Comment (r2k1) [reply e2e/config/azure.go:129]: I avoid including `error` or `failed` in the error message. It just produces a lot of duplication and noise when you wrap errors multiple times:

```
ERROR: error create client: error create VMSS client: error create something: error do something
```
vs

```
ERROR: create client: create VMSS client: create something: do something
```

I don't think it's hard to search them, but it improves the message itself when you surface it (sometimes accidental)

Comment (r2k1) [reply e2e/config/azure.go:129]: It also helps if you ever want to apply a fancy formatting to an error. I think it's a common recommendation to omit the keywords.

File: e2e/scenario_test.go (hunk 17 above)
Comment (r2k1) [reply e2e/scenario_test.go]:
```
type NodeBootstrappingType string

const (
	CustomScripts NodeBootstrappingType = "CustomScripts"
	Scriptless    NodeBootstrappingType = "Scriptless"
)
```

File: e2e/installer_test.go (hunk 21 above)
Comment (r2k1) [reply e2e/installer_test.go:56]: Actually, currently there is a single test only that's using it. Other tests are using binary from the VHD.

I'll optimize it once a usage pattern is more established.

**Review by timmy-wright — APPROVED** [review]

File: self-contained/bootstrap_install.sh (hunk 10 above)
Comment (timmy-wright) [reply self-contained/bootstrap_install.sh]: Yeah, this was going to be part of my next PR. We can either remove it from this one or keep it.

**Review by timmy-wright — APPROVED** [review]

File: pkg/agent/baker.go [hunk 26]
```diff
@@ -1001,6 +979,44 @@ func getContainerServiceFuncMap(config *datamodel.NodeBootstrappingConfiguration
...
+		return true
+	}
+	return cs.Properties.OrchestratorProfile.KubernetesConfig.ContainerRuntimeConfig != nil &&
+		cs.Properties.OrchestratorProfile.KubernetesConfig.ContainerRuntimeConfig[datamodel.ContainerDataDirKey] != ""
+}
+
+func GetKubernetesEndpoint(cs *datamodel.ContainerService) string {
+	if cs.Properties.HostedMasterProfile == nil {
```
[the remark this answers is not in the document]
Comment (timmy-wright) [reply pkg/agent/baker.go:1011]: This is existing code that was extracted to make it reusable and testable.

File: pkg/agent/baker.go [hunk 27]
```diff
@@ -1001,6 +979,44 @@ func getContainerServiceFuncMap(config *datamodel.NodeBootstrappingConfiguration
...
+}
+
+func HasDataDir(config *datamodel.NodeBootstrappingConfiguration) bool {
+	cs := config.ContainerService
+	profile := config.AgentPoolProfile
+	if profile != nil && profile.KubernetesConfig != nil && profile.KubernetesConfig.ContainerRuntimeConfig != nil &&
+		profile.KubernetesConfig.ContainerRuntimeConfig[datamodel.ContainerDataDirKey] != "" {
+		return true
+	}
+	if profile.KubeletDiskType == datamodel.TempDisk {
```
[the remark this answers is not in the document]
Comment (timmy-wright) [reply pkg/agent/baker.go:1003]: This is existing code that was extracted to make it reusable and testable.

**Review by djsly — APPROVED** [review]

File: e2e/config/azure.go (hunk 23 above)
Comment (djsly) [reply e2e/config/azure.go:160]: makes sense if the logger prints the level, this is indeed redundant.
